examples/states_ui/pause.rs

```diff
@@ -0,0 +1,107 @@
+use amethyst::{
+    ecs::Entity,
+    input::{is_close_requested, is_key_down},
+    prelude::*,
+    shrev::EventChannel,
+    ui::{UiCreator, UiEvent, UiEventType, UiFinder},
+    winit::VirtualKeyCode,
+    TransEvent,
+};
+
+use crate::menu::MainMenu;
+
+/// Adapted, originally from amethyst/evoli src/states/pause_menu.rs
+
+#[derive(Default)]
+pub struct PauseMenuState {
+    // button entities are created in on_start() and destroyed in on_stop()
+    resume_button: Option<Entity>,
+    exit_to_main_menu_button: Option<Entity>,
+    exit_button: Option<Entity>,
+    root: Option<Entity>,
+}
+
+// ID's for buttons in the prefab. Required to identify them.
+const RESUME_BUTTON_ID: &str = "resume";
+const EXIT_TO_MAIN_MENU_BUTTON_ID: &str = "exit_to_main_menu";
+const EXIT_BUTTON_ID: &str = "exit";
+
+// load the pause_menu.ron prefab then instantiate it
+// if the "resume" button is clicked, goto MainGameState
+// if the "exit_to_main_menu" button is clicked, remove the pause and main game states and go to MenuState.
+// if the "exit" button is clicked, quit the program.
+impl<'a> SimpleState for PauseMenuState {
+    fn on_start(&mut self, data: StateData<GameData>) {
+        let world = data.world;
+
+        self.root =
+            Some(world.exec(|mut creator: UiCreator<'_>| creator.create("ui/pause_menu.ron", ())));
+    }
+
+    fn on_stop(&mut self, data: StateData<GameData>) {
+        if let Some(root) = self.root {
+            if data.world.delete_entity(root).is_ok() {
+                self.root = None;
+            }
+        }
+        self.resume_button = None;
+        self.exit_to_main_menu_button = None;
+    }
+
+    fn handle_event(&mut self, data: StateData<GameData>, event: StateEvent) -> SimpleTrans {
+        match event {
+            StateEvent::Window(event) => {
+                if is_close_requested(&event) || is_key_down(&event, VirtualKeyCode::Escape) {
+                    Trans::Pop
+                } else {
+                    Trans::None
+                }
+            }
+            StateEvent::Ui(UiEvent {
+                event_type: UiEventType::Click,
+                target,
+            }) => {
+                if Some(target) == self.resume_button {
+                    log::info!("Resuming Game!");
+                    Trans::Pop
+                } else if Some(target) == self.exit_to_main_menu_button {
+                    let mut state_transition_event_channel = data
+                        .world
+                        .write_resource::<EventChannel<TransEvent<GameData, StateEvent>>>();
+
+                    log::info!("Exiting Game, Switching to MainMenu!");
+
+                    // this allows us to first 'Pop' this state, and then exchange whatever was
+                    // below that with a new MainMenu state.
+                    state_transition_event_channel.single_write(Box::new(|| Trans::Pop));
```

Comment:
clever :v:, makes me think that instead of having a `State` stack, we should have a state graph, and transitions to indicate which state to go to instead. (so we don't have to hack around jumping two (or more) levels up).

Reply:
I'm pretty happy with the stack, but an additional graph would be nice. I mean with the stack I can put multiple states on top of each other (since it is much more than just a simple state machine which can ever only be in one state); I don't immediately see how I would do this in the Graph.